File: src/Level/minefield.py
from grid_square import GridSquare

from random import randint

class Minefield:
    """ Represents the Minefield for the current level """
    
    def __init__(self, rows, columns):
        """ Initialize the mine field """
        self.squares = []
        
        for row in range(rows):
            self.squares.append([])
            for column in range(columns):
                self.squares[row].append(GridSquare(row, column))
                
    def scan(self, row, column, drone):
        """ Scan the given Grid Square """
        self.getSquare(row, column).scan(drone)
        
    def defuse(self, row, column, drone):
        """ Defuse the given Grid Square """
        self.getSquare(row, column).defuse(drone)
        
    def defuseCarefully(self, row, column, drone):
        """ Defuse the given Grid Square """
        self.getSquare(row, column).defuseCarefully(drone)
        
    def performEMP(self, row, column, drone):
        """ Performan EMP Blast """
        centerSquare = self.getSquare(row, column)
        for square in self.getAdjacentSquares(centerSquare, includeCenter=True):
            square.hitByEMP(drone)
    
    def addMine(self, mine):
        """ Adds a Mine to the Minefield """
        while True:
            row = randint(0, self.rowCount()-1)
            column = randint(1, self.columnCount()-1)
            square = self.getSquare(row, column)
            if square.mined():
                continue
            else:
                square.setGroundContent(mine)
                mine.gridSquare = square
                break
        
    def getSquare(self, row, column):
        """ Return the square at the location """
        if row < 0 or column < 0 or row >= self.rowCount() or column >= self.columnCount():
            return None
        else:
            return self.squares[row][column]
            
    def getAdjacentSquares(self, centerSquare, includeCenter=False):
        """ Get Squares adjacent to the given square """
        squares = []
        for row in (centerSquare.row-1, centerSquare.row, centerSquare.row+1):
            for column in (centerSquare.column-1, centerSquare.column, centerSquare.column+1):
                square = self.getSquare(row, column)
                if not includeCenter and square is centerSquare:
                    continue
                if square is not None:
                    squares.append(square)
        return squares
    
    def rowCount(self):
        """ Return the Row Count """
        return len(self.squares)
    
    def columnCount(self):
        """ Return the Column Count """
        return len(self.squares[0])
        
    def __repr__(self):
        """ Print the String representation of the Minefield """
        fieldString = ""
        for row in self.squares:
            for column in row:
                fieldString += "|{0}".format(column)
            fieldString += "|\n"
        return fieldString

File: src/android/src/com/kaottessur/pymine/level/Minefield.java
package com.kaottessur.pymine.level;

import java.util.ArrayList;
import java.util.List;

public class Minefield {
	private int rows;
	private int columns;
	
	private List<List<GridSquare>> gridSquares;
	
	
	public Minefield(int rows, int columns) {
		this.rows = rows;
		this.columns = columns;
		
		populateGridSquares();
	}
	
	private void populateGridSquares() {
		gridSquares = new ArrayList<List<GridSquare>>();
		
		for (int row = 0; row < rows; row++) {
			gridSquares.add(new ArrayList<GridSquare>());
			for (int column = 0; column < columns; column++) {
				GridSquare gridSquare = new GridSquare(row, column);
				gridSquares.get(row).add(gridSquare);
			}
		}
	}
	
	public void scan(int row, int column, Drone drone) {
		GridSquare gridSquare = getGridSquareAt(row, column);
		gridSquare.scan(drone);
	}
	
	public void defuse(int row, int column, Drone drone) {
		GridSquare gridSquare = getGridSquareAt(row, column);
		gridSquare.defuse(drone);
	}
	
	public void defuseCarefully(int row, int column, Drone drone) {
		GridSquare gridSquare = getGridSquareAt(row, column);
		gridSquare.defuseCarefully(drone);
	}
	
	public int getRowCount() {
		return rows;
	}
	
	public int getColumnCount() {
		return columns;
	}
	
	public GridSquare getGridSquareAt(int row, int column) {
		try {
			return gridSquares.get(row).get(column);
		} catch(ArrayIndexOutOfBoundsException e) {
			return null;
		} catch(IndexOutOfBoundsException e) {
			return null;
		}
	}
	
	public List<GridSquare> getAdjacentGridSquares(GridSquare centerSquare, boolean includeCenter) {
		List<GridSquare> adjacentGridSquares = new ArrayList<GridSquare>();
		
		for (int row = centerSquare.getRow()-1; row <= centerSquare.getRow()+1; row++) {
			for (int column = centerSquare.getColumn()-1; column <= centerSquare.getColumn()+1; column++) {
				GridSquare gridSquare = getGridSquareAt(row, column);
				
				if (!includeCenter && gridSquare == centerSquare)
					continue;
				
				if (gridSquare != null)
					adjacentGridSquares.add(gridSquare);
			}
		}
		
		return adjacentGridSquares;
	}
	
	public List<GridSquare> getCardinallyAdjacentGridSquares(GridSquare centerSquare) {
		List<GridSquare> adjacentGridSquares = new ArrayList<GridSquare>();
		
		for (int row = centerSquare.getRow()-1; row <= centerSquare.getRow()+1; row+=2) {
			GridSquare gridSquare = getGridSquareAt(row, centerSquare.getColumn());
			if (gridSquare != null)
				adjacentGridSquares.add(gridSquare);
		}
		
		for (int column = centerSquare.getColumn()-1; column <= centerSquare.getColumn()+1; column+=2) {
			GridSquare gridSquare = getGridSquareAt(centerSquare.getRow(), column);
			if (gridSquare != null)
				adjacentGridSquares.add(gridSquare);
		}
		
		return adjacentGridSquares;
	}
}
